bemb/model/bemb.py

```diff
@@ -1298,7 +1316,17 @@ def elbo(self, batch: ChoiceDataset, num_seeds: int = 1) -> torch.Tensor:
         # ==============================================================================================================
         # 1. sample latent variables from their variational distributions.
         # ==============================================================================================================
-        sample_dict = self.sample_coefficient_dictionary(num_seeds)
+        if self.deterministic_variational:
+            num_seeds = 1
+            # Use the means of variational distributions as the sole deterministic MC sample.
+            # NOTE: here we don't need to sample the obs2prior weight H since we only compute the log-likelihood.
+            # TODO: is this correct?
+            sample_dict = dict()
+            for coef_name, coef in self.coef_dict.items():
+                sample_dict[coef_name] = coef.variational_distribution.mean.unsqueeze(
+                    dim=0)  # (1, num_*, dim)
+        else:
+            sample_dict = self.sample_coefficient_dictionary(num_seeds)
```

Comment:
we might want to assert `obs2prior` is False here. If you look into the `sample_coefficient_dictionary` method here:
```
def sample_coefficient_dictionary(self, num_seeds: int) -> Dict[str, torch.Tensor]:
        sample_dict = dict()
        for coef_name, coef in self.coef_dict.items():
            s = coef.rsample(num_seeds)
            if coef.obs2prior:
                # sample both obs2prior weight and realization of variable.
                assert isinstance(s, tuple) and len(s) == 2
                sample_dict[coef_name] = s[0]
                sample_dict[coef_name + '.H'] = s[1]
            else:
                # only sample the realization of variable.
                assert torch.is_tensor(s)
                sample_dict[coef_name] = s
        return sample_dict
```
we sampled the `H` variable in obs2prior term (e.g., $\mu \overset{prior}{\sim} \mathcal{N}(H X^{obs}, \sigma^2 I)$) as well. If we are allowing for deterministic inference with obs2prior (does this make sense at all?), we might want to add the mean of `H` matrices here.

Reply:
Yes I think the right thing to do is to add the mean of H matrices here too. Then we do not have to keep obs2prior = False. I think it does make sense even in that setup, to allow obs2prior. I'm going to try this and get back.